File: Blind150/1.Array&Hash/GO/encode_decode_string.go
package main

import (
	"fmt"
	"strconv"
)

type Solution struct{}

func (s *Solution) Encode(strs []string) string {

	res := ""

	for _, s := range strs {

		res += strconv.Itoa(len(s)) + "C" + s
	}

	return res

}

func (s *Solution) Decode(encoded string) []string {

	res := make([]string, 0)

	i := 0

	for i < len(encoded) {

		j := i

		for string(encoded[j]) != "C" {

			j++
		}

		count, _ := strconv.Atoi(encoded[i:j])

		res = append(res, encoded[j+1:j+1+count])

		i = j + 1 + count

	}
	return res

}
func main() {
	solution := Solution{}

	// Example usage:
	originalStrings := []string{"hello", "world", "golang"}
	encoded := solution.Encode(originalStrings)
	fmt.Println("Encoded:", encoded)

	decoded := solution.Decode(encoded)
	fmt.Println("Decoded:", decoded)
}

// output
// Encoded: 5Chello5Cworld6Cgolang
// Decoded: [hello world golang]

File: Blind150/1.Array&Hash/Python/encode_decode_string.py
# https://neetcode.io/problems/string-encode-and-decode
from typing import List
class Solution:

    def encode(self, strs: List[str]) -> str:

        res = ""
        for s in strs:
            res += str(len(s))+"C"+s
        return res

    def decode(self, s: str) -> List[str]:
        res, i = [], 0


        while i < len(s):

            j = i

            while str(s[j]) != "C":
                j += 1

            count = int(s[i:j])

            res.append(s[j+1: j+1+count])

            i = j+1+count

        return res
